htdocs/bom/tpl/objectline_view.tpl.php

```diff
@@ -77,300 +75,139 @@
+
+		if ($filtertype != 1) {
+			if (getDolGlobalInt('PRODUCT_USE_UNITS')) {
+				$label = measuringUnitString($data->fk_unit, '', '', 1);
+				$column[] = '<td class="linecoluseunit nowrap left">'.
+					(($label !== '') ? $langs->trans($label) : '').'</td>';
+			}
 
-		$sub_bom = new BOM($object->db);
-		if (!empty($obj->fk_bom_child)) {
-			$sub_bom->fetch($obj->fk_bom_child);
-		}
+			$column[] = '<td class="linecolqtyfrozen nowrap right">'.
+				($line->qty_frozen ? yn($data->qty_frozen) : '').'</td>';
 
-		$sub_bom_line = new BOMLine($object->db);
-		$sub_bom_line->fetch($obj->rowid);
+			$column[] = '<td class="linecoldisablestockchange nowrap right">'.
+				($line->disable_stock_change ? yn($data->disable_stock_change) : '').'</td>';
 
-		//If hidden conf is set, we show directly all the sub-BOM lines
-		if (!getDolGlobalString('BOM_SHOW_ALL_BOM_BY_DEFAULT')) {
-			print '<tr style="display:none" class="sub_bom_lines" parentid="'.$line->id.'">';
+			$column[] = '<td class="linecolefficiency nowrap right">'.$data->efficiency.'</td>';
 		} else {
-			print '<tr class="sub_bom_lines" parentid="'.$line->id.'">';
-		}
-
-		// Product OR BOM
-		print '<td style="padding-left: 5%" id="sub_bom_product_'.$sub_bom_line->id.'">';
-		if (!empty($obj->fk_bom_child)) {
-			print $sub_bom_product->getNomUrl(1);
-			print ' '.$langs->trans('or').' ';
-			print $sub_bom->getNomUrl(1);
-		} else {
-			print $sub_bom_product->getNomUrl(1);
-			print '</td>';
+			// Unit
+			$unit = '?';
+			if (!empty($data->fk_unit)) {
+				require_once DOL_DOCUMENT_ROOT.'/core/class/cunits.class.php';
+				$unit = new CUnits($db);
+				$unit->fetch($data->fk_unit);
+				$unit = isset($unit->label) ? "&nbsp;".$langs->trans(ucwords($unit->label))."&nbsp;" : '';
+			}
+			$column[] = '<td class="linecolunit nowrap right">'.$unit.'</td>';
+
+			// Work station
+			if (isModEnabled('workstation')) {
+				$workstation = new Workstation($db);
+				$res = $workstation->fetch($data->fk_default_workstation);
+				$column[] = '<td class="linecolworkstation nowrap right">'.
+					(($res > 0)?$workstation->getNomUrl(1):'none').'</td>';
+			}
 		}
 
-		// Qty
-		$label = $sub_bom_product->getLabelOfUnit('long');
-		if ($sub_bom_line->qty_frozen > 0) {
-			print '<td class="linecolqty nowrap right" id="sub_bom_qty_'.$sub_bom_line->id.'">'.price($sub_bom_line->qty, 0, '', 0, 0).'</td>';
-			if (getDolGlobalString('PRODUCT_USE_UNITS')) {
-				print '<td class="linecoluseunit nowrap left">';
-				if ($label !== '') {
-					print $langs->trans($label);
-				}
-				print '</td>';
+		// Cost
+		$column[] = '<td id="costline_'.$data->id.'" class="linecolcost nowrap right"><span class="amount">'.
+			price(price2num($data->total_cost*$quantity, 'MT')).'</span></td>';
+
+		if ($data->status == 0 && ($object_rights->write) && $action != 'selectlines') {
+			$column[] = '<td class="linecoledit center">'.
+				(($line->info_bits & 2) == 2 || !empty($disableedit)?'':
```

Comment:
If $line is provided from the function that includes the template (tpl), then you can add
```php
'@phan-var-force BOMLine $line';
```
somewhere near the head of the file (BOMLine being replace with the appropriate type of course).

Reply:
Thanks for having a look, I actually copy and pasted that back in but because it is inside the function I renamed $line to $data which in hindsight wasn't the best idea, I have commited a fix which corrects the $line to $data. I did test and it appears to work correctly.